Ability to set discoverable bolt URI explicitly

jakewins opened the pull request:
TL;DR: This makes it possible to override which URI is
published at the HTTP discovery endpoint, letting an administrator
suggest a "bolt+routing://" URI as an alternative to the default
"bolt://" one.

It also expands test coverage for the existing logic to calculate the default.

This has two elements that will make it confusing - the term
"discovery", and the idea of a "public" URI for Bolt.

"Discovery" in Neo could refer both to the discovery *protocol*,
used by causal clustering, or to REST API Discovery, the idea
that the HTTP API describes its own capabilities. This PR is
exclusively about REST/HTTP discovery - it is adding a capability
to the JSON document a user recieves when she does:

    GET http://localhost:7474

The "public" URI portion is more complex, and is the core of what
this PR is mucking about in. Currently, you can find a Bolt URL
to a Neo4j instance in two ways - by asking the HTTP discovery
endpoint, or by asking the routing table procedure. Currently,
both of these methods will give you URLs ultimately originating
from the configuration of the `advertised_address` on the bolt
connector.

However, they serve two different purposes. The routing table is
used *internally* by Bolt Drivers, providing addresses to connect
to exact and specific Neo4j instances.

The Discovery URL is used *externally*, as the root URL to give to
a driver. By connecting via HTTP and retrieving the discovery
document, tools like the browser find out the appropriate URL
to pass when creating a Bolt Driver.

Hence - a sensible "Advertised" URL is always a "bolt://..." URL
pointing to the specific instance being configured. However, a
sensible "Discoverable" URL is sometimes different - specifically,
it may very reasonably be "bolt+routing://path-to-some-load-balancer".

jakewins committed:
Ability to override discoverable bolt URI

This introduces a minor overhaul of the discovery
service, allowing arbitrary URLs to be published to it.
That functionality is leveraged to have the enterprise
edition of the server publish a `bolt+routing` URI.

--

This has two elements that will make it confusing - the term
"discovery", and the idea of a "public" URI for Bolt.

"Discovery" in Neo could refer both to the discovery *protocol*,
used by causal clustering, or to REST API Discovery, the idea
that the HTTP API describes its own capabilities. This PR is
exclusively about REST/HTTP discovery - it is adding a capability
to the JSON document a user recieves when she does:

    GET http://localhost:7474

The "public" URI portion is more complex, and is the core of what
this PR is mucking about in. Currently, you can find a Bolt URL
to a Neo4j instance in two ways - by asking the HTTP discovery
endpoint, or by asking the routing table procedure. Currently,
both of these methods will give you URLs ultimately originating
from the configuration of the `advertised_address` on the bolt
connector.

However, they serve two different purposes. The routing table is
used *internally* by Bolt Drivers, providing addresses to connect
to exact and specific Neo4j instances.

The Discovery URL is used *externally*, as the root URL to give to
a driver. By connecting via HTTP and retrieving the discovery
document, tools like the browser find out the appropriate URL
to pass when creating a Bolt Driver.

Hence - a sensible "Advertised" URL is always a "bolt://..." URL
pointing to the specific instance being configured. However, a
sensible "Discoverable" URL is sometimes different - specifically,
it may very reasonably be "bolt+routing://path-to-some-load-balancer".

jakewins commented:
/cc @jimwebber

technige commented:
@jakewins I've just talked this through with @eastlondoner so I now understand the reasoning behind it. 

Exposing the routing URI makes a lot of sense. I would prefer however to add an additional `bolt+routing` key instead of making the value attached to the `bolt` key switchable. This is because the direct and routing URI forms are not completely semantically interchangeable, meaning that it is perfectly valid for the client to want to select one or the other based on its need. If the server restricts the selection to just one of these, it removes the options available to the client.

I understand that this will require browser changes too but I think it is the correct solution for the discovery service in general.

jimwebber commented on community/server/src/main/java/org/neo4j/server/rest/discovery/DiscoverableBoltURI.java:
```diff
@@ -0,0 +1,88 @@
+
+class DiscoverableBoltURI
+{
+    static String calculateDiscoverableBoltURI( Config config, ConnectorPortRegister connectorPortRegister,
```
Why not URI as the return type?

jakewins replied:
I started with that, but then realized the value eventually needs to become a string when it's passed to `putAbsoluteUri` in the representation, so I changed it to string to be consistent with managementUri and dataUri both being strings. 

TL;DR: I chose an inferior type because I couldn't deal with three variables all having the same suffix but different types.. I'll change it back :)

jakewins commented:
@nigelsmall that makes a lot of sense; and in fairness it may very well be that someone connecting to this instance has a legitimate reason to want to connect directly to it with Bolt. I'll add `bolt+routing` as an optional fourth URL listed

jakewins commented:
@jimwebber @nigelsmall updated to publish a separate `bolt+routing` uri iff the server is enterprise and in non-standalone mode. 

Note two big changes: 

One, this introduces the need to publish different types of URIs depending on the edition. I refactored te discovery service code to use a little registry that tracks the URIs to publish, and modified enterprise edition to add the `bolt+routing` URI if in non-standalone mode with a bolt connector configured.

Two, breaking that registry out of the Jetty code created an annoying dependency on UriInfo, which is something we normally get from Jetty; getting a hold of that before Jetty is booted is a bit clunky. That prompted reviewing what we use it for, and noting there were two cases:

- If the Bolt connector is configured to listen to `localhost`, we modify the discovery URI to be `bolt://<host http is bound to>:<port bolt is bound to>`
- If there is no Bolt connector configured, we publish a discovery URI saying Bolt is listening to `bolt://<host http is bound to>:7687`

I don't think either of these make sense. 

In the first instance, if Bolt is set up to listen to `localhost`, advertising it as listening to another host will cause clients to connect to that other host, which obviously won't work.

In the second instance, if there is no Bolt connector enabled, then there will be no Bolt server listening on the default port - telling users they can find Bolt there will again cause errors.

So: This PR now proposes both of those behaviors are bugs - any client that relies on them will break - and hence it is Good (tm) to change this behavior in the next minor release, 3.4. The behavior implemented in this PR is that if the "advertised" URL for Bolt is configured to "localhost", then that is what the discovery service publishes, and if there is no Bolt connector configured, then there will be no Bolt URL published at the discovery service.

jakewins commented:
@nigelsmall ping, thoughts on this?

MishaDemianenko committed:
Bump asm version to 6.1.1

Upgrade asm to 6.1.1. Fix failing code generation tests on java 10.

MishaDemianenko committed:
Mark all visible for testing stuff with a annotation specific for that instead of comment

Aslo cleanup additional configuration of VisibleForTesting annotation.

MishaDemianenko committed:
Exit on OOM errors during tests.

Exit on OOM during test execution to avoid cases when test can survive
and hide that OOM.

fickludd committed:
Clean up Fingerprint and Monitors around PipeExecutionPlanBuilder

fickludd committed:
Generalize mapping of logical plans.

fickludd committed:
Move PlannerName out of PipeInfo, and clean PipeExecutionBuilderContext

fickludd committed:
Change CommunityPipeBuilder to use semanticTable directly.

fickludd committed:
Move CommunityPipeBuilder to the interpreted-runtime module

fickludd committed:
Rename CommunityPipeBuilder to InterpretedPipeBuilder

fickludd committed:
Remove ReadOnlies attribute: only per-query flag is needed.

fickludd committed:
Remove PipeInfo

Commit:
Adds primitive implementations of DISTINCT

The general purpose DISTINCT implementation uses a
lot of objects to make the output distinct.
This implementation avoids objects and uses primitive
sets to do work, which should be create less memory
pressure and is easier for the CPU to do efficiently.

MishaDemianenko committed:
Unify and extend thread dump utils used for reporting and during testing

Update Verbose timeout to use more tools friendly full thread dump,
unify how we make thread dump between different tools.

SaschaPeukert committed:
Update poms to 3.5

oskarhane committed:
Merge branch '3.4' into 3.5

fickludd committed:
Set the Cypher version to the planner version.

ragadeeshu committed:
Merge branch '3.4' into 3.5

fickludd committed:
Merge remote-tracking branch 'neo4j/3.4' into 3.5

Commit:
Update the Cypher versions to 3.5

lutovich committed:
Merge branch 3.4 into 3.5

Commit:
Merge pull request #11692 from fickludd/3.4-fix-wrong-version-of-compiler-in-plan

Set the Cypher version to the planner version.

Commit:
Merge remote-tracking branch 'upstream/3.4' into 3.5

chrisvest committed:
Merge branch '3.4' into 3.5

Commit:
Merge remote-tracking branch 'upstream/3.4' into 3.5

fickludd committed:
Merge remote-tracking branch 'neo4j/3.4' into 3.5

Commit:
Merge remote-tracking branch 'upstream/3.4' into 3.5

MishaDemianenko committed:
Merge pull request #11543 from MishaDemianenko/3.4-visible-for-testing

Mark all visible for testing stuff with an annotation specific for that instead of comment

Commit:
Merge remote-tracking branch 'upstream/3.4' into 3.5

MishaDemianenko committed:
Merge pull request #11549 from MishaDemianenko/3.4-tests-oom

Exit on OOM errors during tests.

MishaDemianenko committed:
Merge pull request #11538 from MishaDemianenko/3.4-asm-update

Bump asm version to 6.1.1

MishaDemianenko committed:
Merge 3.4 into 3.5

Commit:
Merge pull request #11697 from systay/3.5-cypher-3.5

Update the Cypher versions to 3.5

Commit:
Fixups

ragadeeshu committed:
Merge branch '3.4' into 3.5

fickludd committed:
Merge remote-tracking branch 'neo4j/3.4' into 3.5

fickludd committed:
Forward-merge fixup

henriknyman committed:
Merge branch '3.4' into 3.5

Commit:
Merge remote-tracking branch 'upstream/3.4' into 3.5

Commit:
Merge remote-tracking branch 'upstream/3.4' into 3.5

Commit:
Merge remote-tracking branch 'upstream/3.4' into 3.5

oskarhane committed:
Merge branch '3.4' into 3.5

sherfert committed:
Merge remote-tracking branch 'origin/3.4' into 3.5

RagnarW committed:
Merge remote-tracking branch 'origin/3.4' into 3.5

fickludd committed:
rebase fixup

fickludd committed:
Fix broken test

fickludd committed:
Merge pull request #11602 from fickludd/3.5-baby-steps-towards-cleaner-runtime-interface

Baby steps towards cleaner runtime interface

henriknyman committed:
Merge branch '3.4' into 3.5

martinfurmanski committed:
Merge remote-tracking branch 'origin/3.4' into 3.5

sherfert committed:
Merge remote-tracking branch 'origin/3.4' into 3.5

Commit:
Merge branch '3.4' into 3.5
Merge remote-tracking branch 'upstream/3.5' into 3.5-primitive-distinct

glindroth committed:
Merge branch '3.4' into 3.5

Commit:
More stringent asserts

SaschaPeukert committed:
Merge remote-tracking branch 'origin/3.4' into 3.5

MishaDemianenko committed:
Merge pull request #11675 from MishaDemianenko/3.4-thread-dump-rule

Unify and extend thread dump utils used for reporting and during testing

fickludd committed:
Merge pull request #11713 from systay/3.5-primitive-distinct

Adds primitive implementations of DISTINCT

jakewins commented:
Retargeting for 3.5
